### DISCUSSION TITLE
Each deploy overrides both environments env variables

### DISCUSSION THREAD

**Discussion by gsscoder** [vercel/community#3711] [Bug Report]
### Page to Investigate

https://ucme-cliapp-test.vercel.app/showcase

### Steps to Reproduce

The application has defined the following `next.config.js` file:

```javascript
/* eslint-disable @typescript-eslint/no-var-requires */
/** @type {import('next').NextConfig} */

const withNextBundleAnalyzer = require('@next/bundle-analyzer')({
    enabled: process.env.ANALYZE === 'true',
});

const envPrefix = process.env.VERCEL_ENV === 'preview' ? 'prev_' :
                  process.env.VERCEL_ENV === 'production' ? 'prod_' : '';

module.exports = withNextBundleAnalyzer({
    images: {
        unoptimized: true,
        domains: ['arweave.net'],
        remotePatterns: [
            {
                protocol: 'https',
                hostname: 'nft-preview-media.s3.us-east-1.amazonaws.com',
            },
        ],
    },
    publicRuntimeConfig: {
      baseAppUrl: process.env[`${envPrefix}baseAppUrl`],
      axiosTimeout: parseInt(process.env[`${envPrefix}axiosTimeout`]),
      appiInstrumentationKey: process.env[`${envPrefix}appiInstrumentationKey`],

      b2cSignUpSignInPolicy: process.env[`${envPrefix}b2cSignUpSignInPolicy`],
      b2cInstance: process.env[`${envPrefix}b2cInstance`],
      b2cDomain: process.env[`${envPrefix}b2cDomain`],
      b2cClientId: process.env[`${envPrefix}b2cClientId`],
      b2cRedirectUri: process.env[`${envPrefix}b2cRedirectUri`],

      ucmeApiAssetsServiceUrl: process.env[`${envPrefix}ucmeApiAssetsServiceUrl`],
      ucmeApiCreatorSpaceServiceUrl: process.env[`${envPrefix}ucmeApiCreatorSpaceServiceUrl`],
      ucmeApiIdentityServiceUrl: process.env[`${envPrefix}ucmeApiIdentityServiceUrl`],
      ucmeApiNftAccessServiceUrl: process.env[`${envPrefix}ucmeApiNftAccessServiceUrl`],
      ucmeApiScopes: process.env[`${envPrefix}ucmeApiScopes`],
      ucmeApiErc721JsonUrl: process.env[`${envPrefix}ucmeApiErc721JsonUrl`],

      magicLinkApiKey: process.env[`${envPrefix}magicLinkApiKey`],
      referenceChain: process.env[`${envPrefix}referenceChain`],
      rpcUrl: process.env[`${envPrefix}rpcUrl`],
      environment: process.env[`${envPrefix}environment`]
    }
});
```

Non .env file is committed since from docs I know it's ignored during runtime execution

In the **Vercel dashboard** the variables have all the same name but are prefixed with _prod__ and _prev__ and assigned to Production and Preview environment

Production URL is: [https://ucme-cliapp-test.vercel.app/showcase](https://ucme-cliapp-test.vercel.app/showcase)
Preview URL is: [https://ucme-cliapp-test-giacomostelluti-dev4sidecom.vercel.app/showcase](https://ucme-cliapp-test-giacomostelluti-dev4sidecom.vercel.app/showcase)

Actually the REST call fails cause I intentionally not fixed a CORS issue (which is not a priority now)

Any way if I deploy to Preview with
```sh
$ vercel
```
I get Production and Preview REST calls pointing to Azure services suffixed with **-prd**

And if if I deploy to Production with
```sh
$ vercel --prod
```
I get the both environment pointing to services suffixed with **-dev**

Am I missing something in deployment logic?

Best regards

**Comment by amyegan** [comment]
Hey @gsscoder. It sounds like you have a fairly complex env var management process. I'm not sure I follow what's happening in your setup to cause variables to get the wrong environment suffix when they should be getting a prefix. However, I do think there is an easier way to get different environment variables per environment.

When you [declare an environment variable](https://vercel.com/docs/projects/environment-variables#declare-an-environment-variable) in project settings on Vercel, you can choose which environment can use that value. If you want different values for production, preview, and development, you can add the same variable three times with a different environment and value each time. If you do it that way, then you don't need to worry about setting a prefix or manage different variable names depending on the environment.

**Reply by gsscoder** [reply]
Hello @amyegan thanks for replying

What you suggest is already done. The variable are not only prefixed with `prev_ ` and  `prod_ ` but also assigned to `Preview` and `Production` environments respectively 

In this way I was sure to get an error or a `undefined`/`null` value when the `next.config.js` try to get a not existing variable

I still not understand the behavior of deployment and its malfunction

**Reply by amyegan** [reply]
I'm not able to replicate the error in my test app. Can you share a smaller version of your repo as a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can run to help you to debug it?

**Reply by gsscoder** [reply]
Thanks again @amyegan. We're working to provide what you requested

**Reply by gsscoder** [reply]
Hello @amyegan creating a stripped down version of our repo is unfeasible due to our intense development allocations 😅

Anyway I created a brand new project with same NextJS and React versions: [next-env-test](https://github.com/gsscoder/next-env-test)

Not sure if it malfunctions in the exact same way but this what happens:
- If you deploy **Preview**: **Production** environment is untouched
- If you deploy **Production**: **Preview** environment is _overwritten_ (or it uses wrong variable)

Hope can be of any help to help us :)

**Reply by gsscoder** [reply]
@amyegan any news? btw we've PRO account. is there any preferred way submit this enquiry?
